src/application/application.ts

```diff
@@ -0,0 +1,103 @@
+import { CoreClient } from '../client';
+import { Stream } from '../client/stream';
+import { handleAPIResponse } from '../util/api';
+import { 
+    EventData,
+    ResultData, 
+    ExecuteTaskReply, 
+    StartServiceReply
+} from '../client/core-client';
+
+type Options = {
+    client: CoreClient
+}
+
+class Application {
+    // api gives access to low level gRPC calls
+    api: CoreClient
+
+    private client: CoreClient
+
+    constructor(options: Options){
+        this.client = options.client;
+        this.api = this.client;
+    }
+
+    async whenEvent(event: Event, task: Task): Promise<Stream<EventData>> {
+        await this.startService(event.serviceID);
+        await this.startService(task.serviceID);
+
+        const stream = this.client.listenEvent({
+            serviceID: event.serviceID,
+            eventFilter: event.filter || '*'
+        })
+        stream.on('data', ({ eventKey, eventData }) => {
+            this.executeTask(task, eventKey, eventData);
+        });
+        return stream;
+    }
+
+    async whenResult(result: Result, task: Task): Promise<Stream<ResultData>> {
+        await this.startService(result.serviceID);
+        await this.startService(task.serviceID);
+
+        const stream = this.client.listenResult({
+            serviceID: result.serviceID,
+            taskFilter: result.task || '*',
+            outputFilter: result.output || '*'
+        });
+        stream.on('data', async ({ outputKey, outputData }) => {
+            await this.executeTask(task, outputKey, outputData);
```

Comment:
Why waiting for the Promise here? And not there: https://github.com/mesg-foundation/mesg-js/pull/20/files#diff-bc3351f31e4b582dafa2dc8531df0fa0R35